homeassistant/helpers/device_registry.py

```diff
@@ -0,0 +1,120 @@
+"""Provide a way to connect entities belonging to one device.
+"""
+
+import logging
+import uuid
+
+import attr
+
+from homeassistant.core import callback
+from homeassistant.loader import bind_hass
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_REGISTRY = 'device_registry'
+
+STORAGE_KEY = 'core.device_registry'
+STORAGE_VERSION = 1
+SAVE_DELAY = 10
+
+
+@attr.s(slots=True, frozen=True)
+class DeviceEntry:
+    """Device Registry Entry."""
+
+    identifiers = attr.ib(type=list)
+    manufacturer = attr.ib(type=str)
+    model = attr.ib(type=str)
+    connection = attr.ib(type=list)
+    name = attr.ib(type=str, default=None)
+    sw_version = attr.ib(type=str, default=None)
+    id = attr.ib(type=str, default=attr.Factory(lambda: uuid.uuid4().hex))
+
+
+class DeviceRegistry:
+    """Class to hold a registry of devices."""
+
+    def __init__(self, hass):
+        """Initialize the device registry."""
+        self.hass = hass
+        self.devices = None
+        self._load_task = None
+        self._sched_save = None
+        self._store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+
+    @callback
+    def async_get_device(self, identifiers: str, connections: tuple):
+        """Check if device is registered."""
+        for device in self.devices:
+            if any([identifier in device.identifiers
+                    for identifier in identifiers]) or \
+               any([connection in device.connection
+                    for connection in connections]):
+                return device
+        return None
+
+    @callback
+    def async_get_or_create(self, identifiers, manufacturer, model,
+                            connection, *, name=None, sw_version=None):
+        """Get device. Create if it doesn't exist"""
+        device = self.async_get_device(identifiers, connection)
+        if device is None:
```

Comment:
Use a guard clause.

```python
if device is not None:
    return device
```

Reply:
isn't that what I did? Just inverted

Reply:
Guard clauses should aim to have the last amount of code under the if-statement.

Reply:
When you know you have a value that you're not going to change anymore before returning it from your function ,return it as soon as possible.